overlord/devicestate/devicemgr.go

```diff
@@ -1126,19 +1150,19 @@ func (m *DeviceManager) ensureInstalled() error {
 		return fmt.Errorf("internal error: core device brand and model are set but there is no model assertion")
 	}
 
-	// check if the gadget has an install-device hook
-	var hasInstallDeviceHook bool
-
-	gadgetInfo, err := snapstate.CurrentInfo(m.state, model.Gadget())
+	// check if the gadget has an install-device hook, do this before
+	// we mark ensureInstalledRan as true, as this can fail if no gadget
+	// snap is present
+	hasInstallDeviceHook, err := m.hasInstallDeviceHook(model)
 	if err != nil {
-		return fmt.Errorf("internal error: device is seeded in install mode but has no gadget snap: %v", err)
+		return fmt.Errorf("internal error: %v", err)
 	}
-	hasInstallDeviceHook = (gadgetInfo.Hooks["install-device"] != nil)
 
 	m.ensureInstalledRan = true
 
 	var prev *state.Task
 	setupRunSystem := m.state.NewTask("setup-run-system", i18n.G("Setup system for run mode"))
+	restartSystem := m.state.NewTask("restart-system-to-run-mode", i18n.G("Ensure next boot to run mode"))
```

Comment:
please add a comment that this is chained below, as it's an usual for us to create tasks far from the point they are chained

Reply:
Done!